### ISSUE TITLE
Remove dependency on event bus pattern in MainActivity

### ISSUE THREAD

**Issue by iamareebjamal** [fossasia/open-event-droidgen#2155]
**Actual Behaviour**

EventBus pattern in the app is abused leading to no testability and hence should be removed. The first step would be to remove it from MainActivity where it is majorly used

**Expected Behaviour**

- [ ] Move event handling to a separate component
- [ ] Wire that component with MainActivity through listener
- [ ] Remove event handling in the separate component

This issue may be broken up in several PRs and hence more issues once taken up (i.e. first PR sent), if need be

Parent Issue: #2152

**Comment by dr0pdb** [comment]
@iamareebjamal I would like to work on it.
We do have an open PR regarding creation of ViewModel for this Activity. That should take away some of the complexity from the Activity. Shall i wait for that PR to get updated and merged or start right away ?

**Comment by iamareebjamal** [comment]
No, ViewModel is for MainActivity interaction. This kind of handling is unrelated to view and thus should be broken into its own component. Then probably we can use this component in MainActivtiy's ViewModel

**Comment by dr0pdb** [comment]
@iamareebjamal Can you please explain a bit more about what needs to be done. 
For eg. let's take the simple context of a ```NoInternetEvent``` event. The ```noInternet``` method below subscribes to the ```NoInternetEvent``` inside the MainActivity and calls a helper function ```downloadFailed``` which shows a snackbar.

``` Java
@Subscribe
    public void noInternet(NoInternetEvent event) {
        downloadFailed(null);
    }

```

With what i have understood, You want me to put the ```noInternet``` method inside a new Component and attach a listener to it from the MainActivity. The ```downloadFailed``` method will remain in the MainActivity and will be called from the new Component using that listener. Am i right ? If not, then please explain what needs to be done. Thanks

**Comment by iamareebjamal** [comment]
The problem is deep, the no internet event can be fired from anywhere, so we need to find out from where is it getting fired? Once we have got the event. Let's say it was being fired at the time of JSON downloading, so we need to move it in a component such that it can be encapsulated in one atomic operation. Such as

try {
    downloadFileComponent.startDownloading();
} catch(InternetNotAvailableException e) {
    Show snackbar
}

This is the very problem I was talking about in the issue, where we could have a basic try catch, we are dealing with in form of events which is a very loosely coupled architecture, but the exceptions are tightly coupled with a single method call. This was a basic example of where an event can be turned to an exception. Because it's a final event, no other processing can be done if this event is fired

Another kind of event would be when a JSON completed downloading, this is present in the form of `SpeakersDownloadEvent`, etc. This can be handled with a `onDownloaded` kind of listener from activity to that component. Where the argument of `onDownloaded` will tell which thing was downloaded. Right now, we are handling about 8 of these events, in the form of SpeakerDownloadEvent, SponsorDownloadEvent, and so and so, but we only need one method for it. The consumer of all these events will become one when used in a listener pattern. 

Event Bus was a good choice here because we can create a download thread or service and forget about it and any times the activity is rotated/destroyed, it won't be terminated or restarted because it is detached, and any event on the occasion of completion of download will be triggered independently of the lifecycle of Activity. Thus, it was a very easy choice, but decoupling creates a lot of disadvantage. And we can reduce it to a single listener pattern because now we have configuration surviving ViewModels.

The only places where Event Bus is good is where two components of apps have no idea about each other or there are multiple subscribers of a particular event.
1. Communication between service to activity
2. Notification about a global event acted locally -> Bookmark Changed

Even if the second point is a good pattern for event bus, still, we are striving to remove decoupling of components so that we can increase maintainability
